### DISCUSSION TITLE
Thoughts on Colors

### DISCUSSION THREAD

**Discussion by migueldeicaza** [tui-cs/Terminal.Gui#1107] [Ideas]
Hello,

I was reminded today that our default color scheme works, but sometimes is too hard to spice up the UI in two fronts: customization of colors, and choosing different defaults.

This is not a proposal yet, just a place to share my thinking after years of delaying thinking about the problem. 

On defaults, I think much progress could be made by having a text format to define the colors and allow for new themes to be loaded or changed this way.   We could have a global setting as well as per-view ways of loading new attributes.

I suggest text because it would be simple for end users to customize this as well, something developers can use, without cooking their own.

So three components: parse color definitions, global settings and per view settings would be a starting point. 

Before I talk about color parsing, let us address the other challenge: black and white vs basic color (extended color is not as big of a problem).

Colors are trivial, but our API does not make it easy to use- and this is to some extent done on purpose, because it could break black and white apps.

So the way to solve this is to use a sort of semantic sets of colors, we already have the basic 4 colors that can be used, and in reality, we use three basic sets for black and white (4 attributes for defaults, 4 for popups (dialogs), and 4 for errors - notice that dialog colors and menu colors are just two themes for the same concept: "not the main color scheme, and not an error")

Currently gui.cs uses attributes to encode those and out startup code assigns meaning to some core elements, either colors or black and white attributes so they work together.  Black and white really imposing the limits here.

I think that what we need is to leverage the semantic meaning for those 4 attributes and 3 scenarios and expose those on the color file format.

By using semantic colors for that 4x3 setup, we could hardcode those values for b&w and let users define whatever they want for colors (basic or extended).

So the above could simplify themes, and different color identities.   It does not solve making it easier to use new colors.

Here, I still think that there is value in defining colors semantically, to ensure apps run fine on black and white.  And the api could be one to create a color given specific foreground/background attributes, as well as their intended effect (hot key, normal, disabled).

Will keep this here and digest a more specific proposal, just didn't want to forget this realization.

**Comment by tig** [comment]
Found this article today on blue, which I found interesting:

https://bsago.me/posts/that-annoying-shade-of-blue

This led me to the fact that Windows Terminal is updated it's default color scheme a while back:

https://devblogs.microsoft.com/commandline/updating-the-windows-console-colors/

> If you clean-install a new build of Windows 10 >= 16257, you’ll get the new colors as the default Console scheme.

<img width="356" alt="image" src="https://user-images.githubusercontent.com/585482/190415288-1fafd0c0-3984-4b1e-990d-3d838a23bc43.png">

Related Issues:

* In v2.0 we'll overhaul the color system: https://github.com/gui-cs/Terminal.Gui/issues/457
* TrueColors (dupe of #457?)  https://github.com/gui-cs/Terminal.Gui/issues/48
* Color Schemes: https://github.com/gui-cs/Terminal.Gui/issues/29
* About to be fixed: https://github.com/gui-cs/Terminal.Gui/issues/1800
* https://github.com/gui-cs/Terminal.Gui/issues/1720

**Comment by tig** [comment]
Regarding accessibility in colors: https://www.whocanuse.com/

**Comment by tig** [comment]
Great list of "falsehoods programmers believe about vision":

https://news.ycombinator.com/item?id=32850662

```
One of the worst "ossifications" of non-standards is the notion that colour is three channels of 0..255, one each for red, green, and blue. Combined with the incorrect assumption that the scale is linear[1], and you have a perfect combination for a Dunning-Kruger effect: A bunch of programmers[2] who think they've figured out "graphics", but haven't even begun to understand how any of it works. Not any of: light, display technology, colour standards, or even the human eye.
A short list of falsehoods programmers believe about vision:

- sRGB is the only graphics standard.

- Monitors are actually sRGB and follow that standard.

- Colour intensities are on a linear scale.

- White is 0x00FFFFFF or 255,255,255

- 50% Grey is half of the above: 127,127,127.

- Blending colours is just a matter of linear interpolation.

- Red, green, and blue have a single consistent definition.

- 8 bits is enough for each colour channel.

- Okay, 10 bits is enough.

- 12 bits is enough, surely.

- 0 means black.

- Okay, 16 means black on televisions.

- Fine, black can have different levels in different standards, but it's always clear which level is black.

- At least the white level is consistently defined.

[1] Narrator: it's not even remotely linear.

[2] Including myself until recently.
```

**Comment by GronbergInfo** [comment]
I would like to add my 5 gray rock coins.
I would also love to have an easier configurable way to add theme-ing/own colors thats ouside the few options in the TerminalGuiDesigner provides.
But i also would like to be able to outline/border lines on textboxes, comboboxes and such.
